<h1>Category Analysis of Popular Apps </h1> 

<h2>Goal:</h2>

Do an in-depth study as to what characteristics and attributes can yield the most profit for in ad revenue among Andriod and IOS apps market.

<h2>Source:</h2>

The data set use to analyse trends in both market came from <a href="https://www.kaggle.com/">Kraggle</a>.

<ul>
<li>Google Market: <a href="https://www.kaggle.com/lava18/google-play-store-apps/">Data</a> contains 10,000 records from google play</li>
<li> Apple Market: <a href="https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/">Data</a> contains 7,000 Records from IOS store</li>
</ul>

<H1> Creating Functions:</H1>

I will be creating two functions to help with processing the date.

<ul>
<li>Extract: import data set in a list</li>
<li>Explore Data: print out the content in a list </li>
</ul>





In [109]:
# this allows the user to enter a CSV file name as a parameter to be used in a variable
from csv import reader
def extract_file(file_name):
    open_file=open(file_name)
    read_file=reader(open_file)
    items=list(read_file)
    return items[0],items[1:]

# this function will help explore the data  
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') 

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [110]:
android_header,android = extract_file('googleplaystore.csv')
ios_header, ios = extract_file('AppleStore.csv')


In [111]:
#printing the android data
print(android_header)
print('\n')
explore_data(android, 0, 2, True)


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


In [112]:
print(ios_header)
print('\n')
explore_data(ios, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


<H1> Data Cleaning:</H1>

There is an error within the google data set, which this discussion <a href="https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion/106176" >thread</a> goes into details. 

By reading the discussion, we now know the issue is with the rating column as there should not be a rating higher than 5.

With those details, I  loop through the list to find any rating higher than 5.



In [113]:
rating=[]
at=[]
for count,row in enumerate(android):
    if float(row[2]) > 5: 
        at.append(count)
        rating.append(row[2])
print("The row tha is causing an issue:", at,)
print("The rating for row", at, ":", rating)
print(android[10472])

The row tha is causing an issue: [10472]
The rating for row [10472] : ['19']
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


From the results we can see the error is at row 10472, which is support by the following <a href=" https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion/106176 
">thread information</a>, which state to be at row number 10472 or 10473 (depending on if you have header) to be the error.


We can deal with this error in two way:
<ul>
<li>Delete the entire row with the data.</li>
<li> Changing the value by best guess.</li>
</ul>

Although I disagree with deleting the whole row as data analysts what to preserve as much data as possible and I would change the rating to 1.9 then delete, to show how to manipulate the data. 

In [114]:
#changing 19 to 1.9
android[10472][2]=1.9
print(android[10472])


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', 1.9, '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [115]:
# deleting that row
del android[10472]
print(android[10472])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


<H1>Dealing with Duplicates:</H1>

The thread mentions duplicates in the data sets. 

The first step in removing duplicates is to first identify entries that have duplicates by either id or names. Which in this case will use the name. 


In [116]:
duplicate_apps=[]
unique_apps=[]

for app in android:
    name = app[0]
    if name in unique_apps:# Using the in operator reduce the need for another loop block
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print("Number of duplicate apps:", len(duplicate_apps))


Number of duplicate apps: 1181


The results show there are 1181 entries that are duplicates and to verify my results, I printed one instant of the app with multiple entries. 

In [117]:
for app in android:
    name=app[0]
    if name == duplicate_apps[9]:
        print(app)

['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']


The second steps are to inspect the records to see if there any distinguished attributes.

Upon inspection, we can see all the information is identical with the exception of the number of downloads. In order to clean our data with the most accurate information. We will need to delete all records except the record with the highest downloads. We could have done it by date if there the column was formated in a date-time string which gives us variability.
To remove duplicates and exclude entries with the highest rating requires a two-step process.

<Ol><li>Is to find the highest rating among the duplicates sets.</li>
<li> Is to delete all the duplicates in the primary sets.</li>
</Ol>


In [118]:
# the number entrires we should have at the end of the cleaning process
print('Expected length:', len(android)-len(duplicate_apps))

Expected length: 9659


In [119]:
# first we are finding the higest rating for the each app
reviews_max={}

for app in android:
    name=app[0]
    n_reviews=float(app[3])
    if name in reviews_max and reviews_max[name]<n_reviews:
        reviews_max[name]=n_reviews
    if name not in reviews_max:
        reviews_max[name]=n_reviews
# I should have the same length of 9659
len(reviews_max)
#print(reviews_max)

9659

In [120]:
# We are going to delete all the duplicate entries.
android_clean=[]
already_added=[]

for apps in android:
    name= apps[0]
    n_reviews=float(apps[3])
    #print(n_reviews)
    #print(reviews_max[name])
    if n_reviews == reviews_max[name] and  name not in already_added:
        android_clean.append(apps)
        already_added.append(name)
len(android_clean)
# I should get 9650
        
    

9659

The above code was a two-step process. The first step was to find the highest value of reviews. Then we can compare the name and value in a loop to add only the name and the highest value review to the primary list to delete any duplicates.

<H1> Removing Non-English Apps: </H1>

Upon further research, we found that there are characters that are Non-English in the name column. Which suggests that the application is in the foreign market. 

Since we are the only interesting in the Us market, we will remove those rows. 

First is to identify which app name has character other then English we will build a function to return false if a more then three-letter is grater the 127 otherwise return true.

In [121]:
def is_english(a_string):
    count=0
    for letter in a_string:
        number_code = ord(letter)#ord function covert letter to Ascii code
        #print(number_code)
        if number_code >127 :
            count+=1
    if count > 3:
        return False
    else:
        return True
    
            
    

Test the function is_english

In [122]:
is_english('Instagram')

True

In [123]:
is_english('爱奇艺PPS -《欢乐颂2》电视剧热播')

False

In [124]:
is_english('Docs To Go™ Free Office Suite')

True

In [125]:
is_english('Instachat 😜')

True

The second steps are to use the is_english function and append all true results in a list. 

In [126]:
us_market_android=[]
for apps in android_clean:
    name=apps[0]
    if is_english(name)==1:
        us_market_android.append(apps)
len(us_market_android)


9614

In [127]:
us_market_ios=[]
for apps in ios:
    name=apps[1]
    if is_english(name)==1:
        us_market_ios.append(apps)
len(us_market_ios)

6183

<H1>Isolating the Free Apps</H1>

Now that we have isolated the US market for both platforms, we want to exclude paid applications and keep free applications in our data set, by exploring the data heading we see that price is an index at 7 for android and index 4 at for ios.

In [128]:
print(android_header)
print('\n')

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']




In [129]:
print(ios_header)
print('\n')

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']




In [130]:
# by looping each row and appending price that is zero give us
# free apps in the use market for andriod
us_android_free=[]
for apps in us_market_android:
    price=apps[7]
    if price == '0':
        us_android_free.append(apps)
len(us_android_free)# give us the results


    

8864

In [131]:
# by looping each row and appending price that is zero give us
# free apps in the use market for ios
us_ios_free=[]
for apps in us_market_ios:
    price=float(apps[4])
    if price == 0:
        us_ios_free.append(apps)
len(us_ios_free)# give us the results


3222

<H1> Trend Analyses:</H1>

Our end goal is to build a popular app for both market domain in Google Play and App Store.

<ol>
<li>
 To accomplish this task is to build a simple premise of the ideal and implemented using a little resource, by adding it to Google Play.</li><br>

<li>To monitor the response and should show an upward trend to further develop the applications. 
</li><br>

<li>Provided the feasibility of the application by putting profit benchmarks and when the benchmarks are met to also deploy the application on the ios platform and submitting it to the App store.
</li><br>

</ol>

Using the data set that we have on hand, we will explore and see what columns would be beneficial to generate a frequency table to uncover hidden trends.

We do this by print the head columns of both data sets.




In [132]:
print(android_header)
print('\n')
print(ios_header)
print('\n')

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']




Now that we review the header from both markets we can use category from the google store and prime_genre from the app store as useful matrics in popularity.

We do this by making a frequency table functions, which take the argument of a list and index what column we what a frequency table for. Which would be for Andriod be index 1 and for ios would be index 11.


In [133]:
def freq_table(a_list, a_index):
    return_dic={}
    for apps in a_list:
        name = apps[a_index]
        if name in return_dic:
            return_dic[name]+=1
        else:
            return_dic[name]=1
    total=len(a_list)
    
    for key,vlaue in return_dic.items():
        return_dic[key]=round((vlaue/total)*100,2)
        
    return return_dic

In [134]:
# testing function for android data sets
android_test=freq_table(android,1)
print(android_test)

{'COMICS': 0.55, 'HOUSE_AND_HOME': 0.81, 'SOCIAL': 2.72, 'PHOTOGRAPHY': 3.09, 'FOOD_AND_DRINK': 1.17, 'HEALTH_AND_FITNESS': 3.15, 'VIDEO_PLAYERS': 1.61, 'NEWS_AND_MAGAZINES': 2.61, 'ENTERTAINMENT': 1.37, 'FAMILY': 18.19, 'GAME': 10.55, 'PRODUCTIVITY': 3.91, 'AUTO_AND_VEHICLES': 0.78, 'BEAUTY': 0.49, 'TOOLS': 7.78, 'LIBRARIES_AND_DEMO': 0.78, 'WEATHER': 0.76, 'DATING': 2.16, 'PARENTING': 0.55, 'ART_AND_DESIGN': 0.6, 'SPORTS': 3.54, 'COMMUNICATION': 3.57, 'SHOPPING': 2.4, 'FINANCE': 3.38, 'BUSINESS': 4.24, 'TRAVEL_AND_LOCAL': 2.38, 'PERSONALIZATION': 3.62, 'EVENTS': 0.59, 'EDUCATION': 1.44, 'MEDICAL': 4.27, 'BOOKS_AND_REFERENCE': 2.13, 'LIFESTYLE': 3.52, 'MAPS_AND_NAVIGATION': 1.26}


In [135]:
# testing fuction for ios data sets
ios_test=freq_table(ios,11)
print(ios_test)

{'Shopping': 1.7, 'Navigation': 0.64, 'Photo & Video': 4.85, 'Book': 1.56, 'Reference': 0.89, 'Catalogs': 0.14, 'Food & Drink': 0.88, 'Productivity': 2.47, 'Games': 53.66, 'Medical': 0.32, 'Entertainment': 7.43, 'Health & Fitness': 2.5, 'Education': 6.29, 'News': 1.04, 'Lifestyle': 2.0, 'Finance': 1.45, 'Weather': 1.0, 'Travel': 1.13, 'Business': 0.79, 'Social Networking': 2.32, 'Music': 1.92, 'Utilities': 3.45, 'Sports': 1.58}


In [136]:
# a function that will sort the frequancy table which include my freq_table function
def display_table(data_set, index):
    table = freq_table(data_set, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [137]:
# display_table
display_table(us_ios_free,11)

Games : 58.16
Entertainment : 7.88
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.51
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02
Productivity : 1.74
Lifestyle : 1.58
News : 1.33
Travel : 1.24
Finance : 1.12
Weather : 0.87
Food & Drink : 0.81
Reference : 0.56
Business : 0.53
Book : 0.43
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12


As we can see from the results  there some insight we can see such as:

The most common genre: Games
Second common genre: Entertainment

If we were to do a grouping among genre we can see that most downloads are entertainment-based rather then functional base.

With this data set we can see there is an oversaturation in the entertainment section which would be harder to develop a break out of the app. But what we can focus on education app as it one of the none-Entertainment genres among the top 5 genres. If done with the approach of entertainment we can be a leader in an emerging genre. 


In [138]:
display_table(us_android_free,1)

FAMILY : 18.91
GAME : 9.72
TOOLS : 8.46
BUSINESS : 4.59
LIFESTYLE : 3.9
PRODUCTIVITY : 3.89
FINANCE : 3.7
MEDICAL : 3.53
SPORTS : 3.4
PERSONALIZATION : 3.32
COMMUNICATION : 3.24
HEALTH_AND_FITNESS : 3.08
PHOTOGRAPHY : 2.94
NEWS_AND_MAGAZINES : 2.8
SOCIAL : 2.66
TRAVEL_AND_LOCAL : 2.34
SHOPPING : 2.25
BOOKS_AND_REFERENCE : 2.14
DATING : 1.86
VIDEO_PLAYERS : 1.79
MAPS_AND_NAVIGATION : 1.4
FOOD_AND_DRINK : 1.24
EDUCATION : 1.16
ENTERTAINMENT : 0.96
LIBRARIES_AND_DEMO : 0.94
AUTO_AND_VEHICLES : 0.93
HOUSE_AND_HOME : 0.82
WEATHER : 0.8
EVENTS : 0.71
PARENTING : 0.65
ART_AND_DESIGN : 0.64
COMICS : 0.62
BEAUTY : 0.6


In [139]:
display_table(us_android_free,9)

Tools : 8.45
Entertainment : 6.07
Education : 5.35
Business : 4.59
Productivity : 3.89
Lifestyle : 3.89
Finance : 3.7
Medical : 3.53
Sports : 3.46
Personalization : 3.32
Communication : 3.24
Action : 3.1
Health & Fitness : 3.08
Photography : 2.94
News & Magazines : 2.8
Social : 2.66
Travel & Local : 2.32
Shopping : 2.25
Books & Reference : 2.14
Simulation : 2.04
Dating : 1.86
Arcade : 1.85
Video Players & Editors : 1.77
Casual : 1.76
Maps & Navigation : 1.4
Food & Drink : 1.24
Puzzle : 1.13
Racing : 0.99
Role Playing : 0.94
Libraries & Demo : 0.94
Auto & Vehicles : 0.93
Strategy : 0.91
House & Home : 0.82
Weather : 0.8
Events : 0.71
Adventure : 0.68
Comics : 0.61
Beauty : 0.6
Art & Design : 0.6
Parenting : 0.5
Card : 0.45
Casino : 0.43
Trivia : 0.42
Educational;Education : 0.39
Board : 0.38
Educational : 0.37
Education;Education : 0.34
Word : 0.26
Casual;Pretend Play : 0.24
Music : 0.2
Racing;Action & Adventure : 0.17
Puzzle;Brain Games : 0.17
Entertainment;Music & Video : 0.17
Casual;

Analyzing the frequency table Google play we can see that the top category are Family, Game, Tool Business and Genres would be Tools, Entertainment, education, and business.

So as we compare two data set we see that ios developers are focus on the entertainment while android developers are more balanced in having both practical apps and fun apps.  

We should also include the number of downloads to give better evidence of our analysis. 


To further support of analysis we are going to use the install column from the google play data set but since the app store does not have this column we will use the total number of user ratings as proximation to the total install.


In [140]:
prime_genre =freq_table(us_ios_free,-5)

for genre in prime_genre:
    total=0
    len_genre=0
    for apps in us_ios_free:
        genre_app=apps[-5]
        if genre_app==genre:
            total+=float(apps[5])
            len_genre+=1
    avg=total/len_genre
    print(genre,":",round(avg)) #sorted(table_display, reverse = True)
    #print(genre)

Shopping : 26920
Navigation : 86090
Photo & Video : 28442
Book : 39758
Reference : 74942
Catalogs : 4004
Food & Drink : 33334
Productivity : 21028
Games : 22789
Medical : 612
Entertainment : 14030
Health & Fitness : 23298
Education : 7004
News : 21248
Lifestyle : 16486
Finance : 31468
Weather : 52280
Travel : 28244
Business : 7491
Social Networking : 71548
Music : 57327
Utilities : 18684
Sports : 23009


Further diving into the data we do see that an app profile such as reference is the top applications with the most downloads, followed by weather, lifestyle. 
Which now we can focus on low app development in the market place with a high download count.


In [141]:
google_cat=freq_table(us_android_free,1)
for category in google_cat:
    total=0
    len_category=0
    for apps in us_android_free:
        category_app=apps[1]
        if category_app==category:
            n_install=apps[5]
            n_install =n_install.replace('+','')
            n_install= n_install.replace(',','')
            total+=float(n_install)
            len_category+=1
    avg=total/len_category
    print(category,":",round(avg))

COMICS : 817657
HOUSE_AND_HOME : 1331541
SOCIAL : 23253652
PHOTOGRAPHY : 17840110
FOOD_AND_DRINK : 1924898
HEALTH_AND_FITNESS : 4188822
VIDEO_PLAYERS : 24727872
NEWS_AND_MAGAZINES : 9549178
ENTERTAINMENT : 11640706
FAMILY : 3695642
GAME : 15588016
PRODUCTIVITY : 16787331
AUTO_AND_VEHICLES : 647318
BEAUTY : 513152
TOOLS : 10801391
LIBRARIES_AND_DEMO : 638504
WEATHER : 5074486
DATING : 854029
PARENTING : 542604
ART_AND_DESIGN : 1986335
SPORTS : 3638640
COMMUNICATION : 38456119
SHOPPING : 7036877
FINANCE : 1387692
BUSINESS : 1712290
TRAVEL_AND_LOCAL : 13984078
PERSONALIZATION : 5201483
EVENTS : 253542
EDUCATION : 1833495
MEDICAL : 120551
BOOKS_AND_REFERENCE : 8767812
LIFESTYLE : 1437816
MAPS_AND_NAVIGATION : 4056942


Looking at both data sets we can see that there room to grow in the education application for both google and ios. 